File: 2021/d1/rust/src/main.rs
use std::path::Path;
use std::fs::File;
use std::io::Read;

fn count(lines: &Vec<usize>, window_size: usize) -> usize {
    let mut count = 0;

    for window in lines.windows(window_size + 1) {
        let mut res = window.windows(window_size).map(|w| w.into_iter().sum::<usize>());
        let first = res.next().unwrap();
        let second = res.next().unwrap();

        if second > first {
            count += 1;
        }
    }

    count
}

fn main() {
    let mut args = std::env::args();

    let input_filename = args.nth(1).unwrap_or_else(|| {
        eprintln!("No input file specified");
        eprintln!("Usage: ./run <input filename>");
        std::process::exit(1);
    });

    let mut input_file = File::open(Path::new(&input_filename)).expect("failed to open input file");
    let mut input = String::new();
    input_file.read_to_string(&mut input).expect("failed to read from file");

    let lines = input.lines()
        .map(|l| l.parse().expect("failed to parse int from line"))
        .collect();

    println!("part 1: {}", count(&lines, 1));
    println!("part 2: {}", count(&lines, 3));
}

File: 2021/d1/python/main.py
import sys

def increases(lines, window_size):
    count = 0
    i = 0

    while i < len(lines) - window_size:
        if lines[i] < lines[i + window_size]:
            count += 1
        i += 1

    return count

if __name__ == '__main__':
    if len(sys.argv) < 2:
        raise Exception("filename not speciifed, specify filename as first argument")

    filename = sys.argv[1]
    lines = None

    with open(filename) as f:
        lines = [int(line.strip()) for line in f]

    if not lines:
        raise Exception("no lines found")

    print("part 1: " + str(increases(lines, 1)))
    print("part 2: " + str(increases(lines, 3)))
